File: Problem/Medium/1-100/98. Validate Binary Search Tree/a.py
import math


class TreeNode:
    def __init__(self, val, left, right):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root):
        def validate(node, low, high):
            if not node:
                return True

            if not (low < node.val < high):
                return False

            return validate(node.left, low, node.val) and validate(node.right, node.val, high)

        return validate(root, -math.inf, math.inf)

File: Problem/Medium/1-100/98. Validate Binary Search Tree/_a.java
class Solution {
	public boolean isValidBST(TreeNode root) {
		return validate(root, Long.MIN_VALUE, Long.MAX_VALUE);
	}

	private boolean validate(TreeNode node, long low, long high) {
		if (node == null)
			return true;

		if (node.val <= low || high <= node.val)
			return false;

		return validate(node.left, low, node.val) && validate(node.right, node.val, high);
	}
}
